File: LList-cycle2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# This is a HashMap solution
# Space complexity is O(n)
# Time complexity is O(n)


class Solution:
    def detectCycle(self, head):
        if head is None:
            return None
        cur = head
        hmaps = {}
        while cur and cur.next:
            if cur not in hmaps:
                hmaps[cur] = cur
            else:
                return hmaps[cur]
            cur = cur.next

        return cur.next

File: LListCycle-2.cpp
// Time Complexity :O(n) where n in the number of nodes
// Space Complexity : O(1)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(!head) return nullptr;
        set<ListNode *> hSet;
        ListNode *rPtr = head;
        while(rPtr){
            if(hSet.find(rPtr) == hSet.end())
                hSet.insert(rPtr);
            else
                return rPtr;
            rPtr = rPtr->next;
        }
        return nullptr;
    }
};